src/client.c

```diff
@@ -1096,8 +1188,62 @@ bool client_proto(SBuf *sbuf, SBufEvent evtype, struct MBuf *data)
 		/* client is not interested in it */
 		break;
 	case SBUF_EV_PKT_CALLBACK:
-		/* unused ATM */
+	{
+		bool first = false;
+		if (client->packet_cb_state.pkt_header.type == 0) {
+			first = true;
+			if (!get_header(data, &client->packet_cb_state.pkt_header)) {
+				char hex[8*2 + 1];
+				disconnect_client(client, true, "bad packet header: '%s'",
+								  hdr2hex(data, hex, sizeof(hex)));
+				return false;
+			}
+			mbuf_rewind_reader(data);
+		}
+
+		switch (client->packet_cb_state.flag) {
+			case CB_WANT_COMPLETE_PACKET:
+				if (first) {
+					slog_debug(client, "buffering complete packet, pkt='%c' len=%d incomplete=%s available=%d",
+							   pkt_desc(&client->packet_cb_state.pkt_header),
+							   client->packet_cb_state.pkt_header.len,
+							   incomplete_pkt(&client->packet_cb_state.pkt_header) ? "true" : "false",
+							   mbuf_avail_for_read(data));
+					mbuf_init_dynamic(&client->packet_cb_state.complete_pkt);
+					if (!mbuf_make_room(&client->packet_cb_state.complete_pkt, client->packet_cb_state.pkt_header.len))
+						return false;
+				}
+
+				if (!mbuf_write_raw_mbuf(&client->packet_cb_state.complete_pkt, data))
+					return false;
+
+				if (sbuf->pkt_remain == mbuf_avail_for_read(data)) {
+					if (!get_header(&client->packet_cb_state.complete_pkt, &pkt)) {
+						char hex[8*2 + 1];
+						disconnect_client(client, true, "bad packet header: '%s'",
+										  hdr2hex(data, hex, sizeof(hex)));
+						return false;
+					}
+					slog_debug(client, "complete packet buffered, pkt='%c' len=%d incomplete=%s", pkt_desc(&pkt), pkt.len, incomplete_pkt(&pkt) ? "true" : "false");
+
+					client->packet_cb_state.flag = CB_HANDLE_COMPLETE_PACKET;
+					res = handle_client_work(client, &pkt);
+
+					mbuf_free(&client->packet_cb_state.complete_pkt);
+					free_header(&client->packet_cb_state.pkt_header);
+				} else {
+					/*
+					 * We wrote the partial packet to our temporary buffer. So
+					 * we "handled" it and want to receive more data.
+					 */
+					res = true;
```

Comment:
@sboschman regarding this request of yours:

> Which brings me right to an item on my wish list, be able to handle especially the parse packet in a sort of streaming manner. I.e. let the client sent the sql query, process bytes with a reasonable small pkt_buf, gather the entire sql query in pgbouncer, do the processing, sent out a modified (parse) packet to the server using a smaller sending buffer. Having to size pkt_buf according to the largest query being sent is not very user friendly.

I pushed a commit which fixes the issue with needing to set pkt_buf to a high value. It does so by essentially doing the two following things:
1. setting `res` to true in this line
2. Setting CB_WANT_COMPLETE_PACKET when the inspect function fails instead of returning `false`. It's possible for the inspection to fail because not the whole packet is available.

Reply:
Nice!

If I'm understanding the code correctly the parse and bind packets are completely buffered in pgbouncer. The parse packet has to be buffered completely as we have to calculate a fingerprint for the named prepared statement. This is not true for the bind packet. In case of a bind packet we only need the first few bytes, so we can modify the name of the source prepared statement. As the bind param values can be many bytes, it might be worth it to not fully buffer a bind packet in pgbouncer.

For example with serverside prepared statements in pgjdbc driver we do not have much control over which queries are  prepared by the driver. Insert/update queries can also end up being prepared by the driver. Some bind params might contain considerable amounts of text/blob bytes.

Reply:
> If I'm understanding the code correctly the parse and bind packets are completely buffered in pgbouncer. 

Yes, correct. (except if they are using unnamed prepared statements)

> As the bind param values can be many bytes, it might be worth it to not fully buffer a bind packet in pgbouncer.

That's a good point, and I agree it would be good to that. But I'm not sure it's super easy to change it so that it does that. I'm leaning towards fully buffering the packet for the first version of this feature.

Reply:
Okay, it wasn't very easy. But I got it to work.